PROTOCOL.md

```diff
@@ -0,0 +1,193 @@
+
+### Requests
+
+**Request ::= Request-Headers Request-Body**
+
+Twirp always uses HTTP POST method to send requests, because it
+closely matches the semantics of RPC methods.
+
+The **Request-Headers** are normal HTTP headers. The Twirp wire
+protocol uses the following headers.
+
+* **Authorization** header is often used to pass user credentials
+  from the client to the server, such as OAuth access token or
+  JWT token.
+
+* **Content-Type** header indicates the proto message encoding, which
+  should be one of "application/x-protobuf", "application/json". The
+  server uses this value to decide how to parse the request body,
+  and encode the response body.
+
+* **User-Agent** header indicates the client application and its
+  runtime environment. While the server should not use this
+  information for request processing, this header is heavily used
+  for analytics and troubleshooting purposes.
+
+* **RPC-Timeout** header indicates the client-specified request
+  timeout in seconds, such as "10". If **RPC-Timeout** is omitted, the
+  server should use a pre-configured timeout value, by default it
+  should be 10 seconds.
+
+The **Request-Body** is the encoded request message, contained in the
+HTTP request body. The encoding is specified by the `Content-Type`
+header.
+
+### Responses
+
+**Response ::= Response-Headers Response-Body**
+
+The **Response-Headers** are just normal HTTP response headers. The
+Twirp wire protocol uses the following headers.
+
+* **Content-Type** The value should be either "application/x-protobuf"
+  or "application/json" to indicate the encoding of the response
+  message. It must match the "Content-Type" header in the request.
+
+The **Request-Body** is the encoded response message contained in the
+HTTP response body. The encoding is specified by the `Content-Type`
+header.
+
+## Example
+
+The following example shows a simple Echo API definition and its
+corresponding wire payloads.
+
+The example assumes the server base URL is "https://example.com".
+
+```proto
+package twirp;
+
+service Echo {
+  rpc Hello(HelloRequest) returns (HelloResponse);
+}
+
+message HelloRequest {
+  string message;
+}
+
+message HelloResponse {
+  string message;
+}
+```
+
+**Proto Request**
+
+```
+POST /twirp.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/x-protobuf
+Content-Length: 15
+
+<encoded HelloRequest>
+```
+
+**JSON Request**
+
+```
+POST /twirp.Echo/Hello HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+**Proto Response**
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/x-protobuf
+Content-Length: 15
+
+<encoded HelloResponse>
+```
+
+**JSON Response**
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Content-Length: 27
+
+{"message":"Hello, World!"}
+```
+
+## Errors
+
+If an error occurs when the server processes a request, the server
+must return an error payload as the response message, and correctly
```

Comment:
I think we should describe Twirp's error payload in detail here.

Twirp error responses are always JSON-encoded (*regardless* of the request's Content-Type), with a corresponding `Content-Type: application/json` header. This ensures that they are human-readable in any setting.

They are a JSON object with three keys:

 - **code**: One of the Twirp error codes as a string.
 - **msg**: A human-readable message describing the error as a string.
 - **meta**: An object with string keys and values holding arbitrary additional metadata describing the error.

For example:
```json
{
    "code": "permission_denied",
    "msg": "thou shall not pass",
    "meta": {
        "target": "Balrog"
    }
}
```

See https://github.com/twitchtv/twirp/wiki/Errors.